module1/lessons/ruby_object_model_poly.md

```diff
@@ -0,0 +1,220 @@
+---
+title: Object-Oriented Programming
+length: 60min
+layout: page
+---
+
+## Learning Goals
+
+* Understand and explain the Look Up Chain and how it relates to the principle of Polymorphism in Object-Oriented Programming.
+* Explore how Ruby's object model facilitates this principle.
+
+## Vocabulary
+
+* Inheritance
+* Polymorphism
+* Superclass
+* Module
+* Object Model
+* Look Up Chain
+
+## WarmUp
+
+* What is class inheritance and how is it implemented in Ruby?
+* What is a module in Ruby and how is it implemented?
+* How do you know what variables, methods, and classes you have available at any given time in Ruby?
+
+# The Ruby Object Model
+
+## Investigative Methods
+
+These three methods can help you investigate the relationships between classes and modules in Ruby. All methods are run on the class (i.e., `String`, `Hash`):
+
+* `.ancestors`: lists all classes along the inheritance chain, and any modules included by those classes.
+* `.included_modules`: returns a list of all modules included by any class along the inheritance chain.
+* `.superclass`: returns the superclass of the class.
+
+## Mapping Ruby's Object Model
+
+### Superclasses
+
+We are going to create a `Dog` class that inherits from another class called `Animal`, which includes a module called `AnimalBehavior`:
+
+```ruby
+# dog.rb
+require './animal'
+
+class Dog < Animal
+end
+
+dog = Dog.new
+require 'pry'; binding.pry
+```
+
+```ruby
+# animal.rb
+require './animal_behavior'
+
+class Animal
+  include AnimalBehavior
+end
+```
+
+```ruby
+# animal_behavior.rb
+module AnimalBehavior
+end
+```
+
+Notice that we haven't actually included any methods in these classes/modules. We don't need them to map Ruby's Object Model.
+
+When we run the `dog.rb` file, we create a new instance of `Dog` and then hit a pry. Using the investigative methods we defined above, we can learn about that dog instance's ancestors:
+
+```ruby
+pry(main)> dog
+# => #<Dog:0x007ff414932eb0>
+
+pry(main)> dog.class
+# => Dog
+
+pry(main)> Dog.ancestors
+# => [Dog, Animal, AnimalBehavior, Object, PP::ObjectMixin, Kernel, BasicObject]
+
+pry(main)> Dog.superclass
+# => Animal
+
+pry(main)> Animal.superclass
+# => Object
+
+pry(main)> Object.superclass
+# => BasicObject
+
+pry(main)> BasicObject.superclass
+# => nil
+```
+
+Calling `.class` on the Dog object leads us to the `Dog` class. Calling `.superclass` on the `Dog` class leads us to `Object`, and calling `superclass` on `Object` leads us to `BasicObject`. `BasicObject` has no superclass, so the inheritance chain ends there. We can summarize this information in a diagram:
+
+![Dog Inheritance](https://i.imgur.com/6IwoHvk.png)
```

Comment:
Consider pulling this image into the repo as an asset.

Reply:
Good call--done!